### PR TITLE
[Merged by Bors] - feat: add the `cdot` linter

### PR DESCRIPTION
**Pull request by adomani**
The `cdot` linter is a syntax linter that flags uses of the "cdot" `·` that are achieved by typing a character different from `·`.

For instance, a "plain" dot `.` is allowed syntax, but is flagged by the linter.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/Lint.lean [hunk 1]
```diff
@@ -158,4 +158,66 @@ initialize addLinter missingEndLinter
 
 end MissingEnd
 
+/-!
+# The `cdot` linter
+
+The `cdot` linter is a syntax-linter that flags uses of the "cdot" `·` that are achieved
+by typing a character different form `·`.
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:165]:
```suggestion
by typing a character different from `·`.

File: Mathlib/Tactic/Linter/Lint.lean [hunk 2]
```diff
@@ -158,4 +158,66 @@ initialize addLinter missingEndLinter
...
+      | ``Lean.Parser.Term.cdot => dargs.push stx
+      | ``cdotTk => dargs.push stx
+      | _ =>  dargs
+  |_ => #[]
+
+/-- `unwanted.cDot stx` returns an array of syntax atoms corresponding to `cdot`s that are not
+written with the character `·`.
+This is precisely what the `cdot` linter flags.
+-/
+def unwanted.cDot (stx : Syntax) : Array Syntax :=
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:201]: Given this is the first linter for unwanted syntax, what do you think about naming this "unwanted_cdot" instead?

File: Mathlib/Tactic/Linter/Lint.lean [hunk 3]
```diff
@@ -158,4 +158,66 @@ initialize addLinter missingEndLinter
...
+the character `·`. -/
+def isCDot? : Syntax → Bool
+  | .node _ ``cdotTk #[.node _ `patternIgnore #[.node _ _ #[.atom _ v]]] => v == "·"
+  | .node _ ``Lean.Parser.Term.cdot #[.atom _ v] => v == "·"
+  | _ => false
+
+/--
+`findDot stx` extracts from `stx` the syntax nodes of `kind` `Lean.Parser.Term.cdot` or `cdotTk`. -/
+partial
+def findDot : Syntax → Array Syntax
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:188]: Can you call this something with "cdot", please?
(I'm unsure how to capitalise names with cdot in it. I may give that bikeshed more thought after lunch.)

File: test/Lint.lean [hunk 4]
```diff
@@ -57,3 +57,23 @@ set_option linter.dupNamespace true in
...
+
+set_option linter.cdot false in
+/--
+warning: Please, use '·' (typed as `\·`) instead of '.' as 'cdot'.
+note: this linter can be disabled with `set_option linter.cdot false`
+---
+warning: Please, use '·' (typed as `\·`) instead of '.' as 'cdot'.
+note: this linter can be disabled with `set_option linter.cdot false`
+---
+warning: Please, use '·' (typed as `\·`) instead of '.' as 'cdot'.
```
Comment (grunweg) [inline test/Lint.lean:69]: Would you like to test a second kind of wrong dot as well, e.g. change of the wrong dots? (I was wondering if the wrong dot was hardcoded - the answer is "clearly not", but a test could be nice.)

File: Mathlib/Tactic/Linter/Lint.lean [hunk 5]
```diff
@@ -158,4 +158,66 @@ initialize addLinter missingEndLinter
 
 end MissingEnd
 
+/-!
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/Lint.lean:161]: This is starting a bikeshed, but: intuitively, I would rather add this to `Linters/Style`. (This linter is purely about coding style, and not at all relevant to correctness.) What do you think?

I agree this starts a broader conversation how to structure the `Linters` directory: would we rather have separate files for individual linters, or are we fine grouping linters if each is reasonably short? (In any case, having an overview of all linters in some doc-string would be nice --- but that's definitely out of scope for this PR.)

**Review by grunweg — COMMENTED** [review]
Looks great in general! Here's a first round of small comments: I'll eat now, and will return to this afterwards.

File: Mathlib/Tactic/Linter/Lint.lean (hunk 2 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Lint.lean]: Renamed it to `unwanted_cdot`.

File: test/Lint.lean (hunk 4 above)
Comment (adomani) [reply test/Lint.lean:69]: I added another test, with `(. + ·)`, swapping the positions of `·` and `.`.

File: Mathlib/Tactic/Linter/Lint.lean (hunk 3 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Lint.lean]: Renamed to `findCDot`.  My instinct would be to use the same capitalization for `C` and `D`, like `ZMod`, `EReal`, `PUnit`,...

File: Mathlib/Tactic/Linter/Lint.lean (hunk 5 above)
Comment (adomani) [reply Mathlib/Tactic/Linter/Lint.lean:161]: I agree that this is more "stylistic" than "correctness", but the other linters also do not really check correctness.

Anyway, I am happy if this linter goes somewhere else, but, I also agree that this is a discussion that should happen elsewhere!

File: Mathlib/Tactic/Linter/Lint.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean]: Works for me!

File: Mathlib/Tactic/Linter/Lint.lean (hunk 5 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Lint.lean:161]: So I'll start a zulip discussion/propose a re-organisation in a PR? Sounds good to me.

File: test/Lint.lean (hunk 4 above)
Comment (grunweg) [reply test/Lint.lean:69]: Thanks! I had meant something else: are there other characters which are treated like a cdot, but are also bad.
Reading `Init/NotationExtra.lean` in core, the current answer is no - so this is fine as-is.

**Review by grunweg — COMMENTED** [review]
Thanks for writing this! I just pushed a commit with tiny style-only tweaks directly.
Looks good to me now.
maintainer merge

**Pull request closed without merging**